### DISCUSSION TITLE
Passing parameters values from anylogic to python without defining the variable in python

### DISCUSSION THREAD

**Discussion by BLindeboom27** [the-anylogic-company/AnyLogic-Pypeline#22] [Q&A] [answered]
Hello,

I am experiencing some difficulties with sending parameter values to python in order to instigate the algorithm. The names within "" are the parameters of the algorithm I would like to have anylogic send to python as I am trying to integrate sliders. 
The following code does not seem to work:

pyCommunicator.run(String.format("generations = " + ParGenerations));
pyCommunicator.run(String.format("population_size  = " + ParPopulationSize));
pyCommunicator.run(String.format("elite = " + ParElite));
pyCommunicator.run(String.format("mutation_rate = " + ParMutationRate));
pyCommunicator.run(String.format("penalty_value = " + ParPenaltyValue));

pyCommunicator.run("import MDHFVRP_postNL_v3");

I have also tried:
String generations = "generations = " + ParGenerations;
String population = "population_size  = " + ParPopulationSize;
String mutation = "mutation_rate = " + ParMutationRate;
String elite = "elite = " + ParElite;
String penalty = "penalty_value = " + ParPenaltyValue;

pyCommunicator.runResults("import MDHFVRP_postNL_v3.py", generations, population, mutation, elite, penalty);

Somebody an idea on how to solve this?

**Comment by t-wolfeadam** [comment, marked as answer]
Hi! It's hard to know what _exactly_ "does not seem to work" without knowing what problem is being encountered or how your Python script is setup.

Am I correct in assuming your your `MDHFVRP_postNL_v3` script uses variables like `generations` and `population_size`, and that you're trying to define them first and then import the file?

**Reply by BLindeboom27** [reply]
Hi!
Thanks for the quick response. Yes, you are correct. 

The idea is to have a hybrid simulation optimization model using a genetic algorithm to solve the vehicle routing problem with multiple depots and a heterogeneous fleet. 

To initialize the genetic algorithm I would like to use the python communicator to send the parameters (`generations, elite, population_size, mutation_rate and penalty_value`) of the genetic algorithm and the properties of the fleet (using agents with connected databases) to the python script `MDHFVRP_postNL_v3.py`. This is all prior to running the actual algorithm, which is instigated via a button. 

With the coding above I receive the error that the parameters are undefined in the python script: `java.lang.RuntimeException: root.pyCommunicator:
NameError("name 'population_size' is not defined"`

In python I have not defined anything with respect to the Anylogic coding, I assume this has to be the case?

**Reply by t-wolfeadam** [reply]
Thanks for the explanation - and interesting idea! 

Based on the error and your code, I'm guessing your `MDHFVRP_postNL_v3` script has its logic executed immediately and references externally defined variables. Afaik, this isn't possible in Python - let alone with Pypeline - due to the way that importing a script works. As an example, here's the same problem in pure Python: [image](https://user-images.githubusercontent.com/40703071/141512792-6f2c6729-deef-448f-9e19-93b2164216e4.png).

In short, since `MDHFVRP_postNL_v3.py` serves as a script you import for your logic, here's two simple ways to refactor:

1. Wrap the logic inside a function with arguments (e.g., generations, population_size, ...) and return the outputs. This is the simplest and most "Pythonic" solution (I recommend it). 
2. If you want to interact manually with variables, first define all input and output variables at the start of `MDHFVRP_postNL_v3.py`. You still need to wrap its logic in a function, but rather than have inputs/outputs, use the `global` keyword at the top for all the variables.

The following is an example with _both_ of these solutions demonstrated: [Model167.zip](https://github.com/t-wolfeadam/AnyLogic-Pypeline/files/7529887/Model167.zip)

**Reply by BLindeboom27** [reply]
Dear Adam,

I now see that I did not thank you at all for your help. I have been successful in processing the functions. 
My sincere gratitude for aiding me!

**Reply by t-wolfeadam** [reply]
My pleasure! Glad to hear it was successful 😄

**Comment by BLindeboom27** [comment]
Hi,

There have been some design changes and I am encountering issues again that I have not been able to resolve myself.  The solutions you provide are applicable to functions with single parameters, however, I have defined multiple functions that use different parameters but the value of the parameters are equal over the different functions. 

For the genetic algorithm, I first have to define functions such as evaluate_capacity and the breeding of new solutions. Here the distance matrix is included also. To retrieve the result I have created a function based on 21 parameters. I want to pass the distance matrix only in this function (if possible even). In the pipeline examples,  the following code is used for a function based on two parameters

 `pyCom.run(
	"import random", // used in the Order agent
	"from tsp_solver import FacilityOrderSolver",
	String.format( // pass in distance matrix and home index
		"solver = FacilityOrderSolver(%s, %s)",
		pyCom.toJson(buildDistanceMatrix()),
		home_index
	)
);`
Is it possible to define only 1 parameter whilst not sending values for the rest? Particularly if the size of the parameters is different and if the code is functional based rather than object-oriented (via the class). 

The following solution I have tried but without success obviously: 
`String distance_matrix = pyCommunicator.toJson(BuildDistanceMatrix()));
pyCommunicator.run("GA_PostNL.GAresult(%s, %s, %s, %s"), distance_matrix);`

Another solution that i see is to create a function for the distance matrix and return the distance matrix as a variable in python. However, here I seem to make some syntax mistake. The solution is lost to me however. 
`VarDepotResult = pyCommunicator.runResults(int[].class, "GA_PostNL.depots", "GA_PostNL.distance_matrix = %s", pyCommunicator.toJson(BuildDistanceMatrix()));
`

Is it possible?

Could you please let me know the possibilities/

Bas


![GAresultfunction](https://user-images.githubusercontent.com/94119326/144867605-5cc08ec4-6c85-458a-bddf-b766d62eb7e6.png)

**Reply by t-wolfeadam** [reply]
While I seem to understand what you're asking, one reason why your provided example call to the "GAresult" function may not have worked is a misplaced parenthesis. I believe the close-parentheses after the double quotation mark should be placed before it: 

```java
pyCommunicator.run("GA_PostNL.GAresult(%s, %s, %s, %s"), distance_matrix);` 
//                                                    ^ here
```

Another reason it does not work is because you put multiple "%s" placeholders but only define one value. If you don't want those to be passed as a Java variable, then you can simply hardcode the values - this can be a valid Python value or a variable in your Python environment. 

In the following example, I reference a made-up function in your "GA_PostNL" script that I called `dummy`. It uses a mix of a Java-based variable, a new Python variable, a made-up Python variable in your "GA_PostNL" script and a static value:

```java
pyCommunicator.run("my_var1 = 3.14");
pyCommunicator.run("GA_PostNL.dummy(%s, my_var1, GA_PostNL.default_settings, True)", distance_matrix);
```

Alternatively, set default values for these in your Python script by converting them to keyword arguments (though you'll need to refactor to place them after any other positional/non-keyword arguments). 

-----

> `VarDepotResult = pyCommunicator.runResults(int[].class, "GA_PostNL.depots", "GA_PostNL.distance_matrix = %s", pyCommunicator.toJson(BuildDistanceMatrix()));` 

Regarding this other code example, there's a few problems: 

1. The Java variable needs to have its type defined (i.e., `int[] VarDepotResult = ...`)
2. The last two arguments are assigning a variable - this is the wrong use for `runResults`; you need to use a separate `run` call for this. I.e., the call to `runResults` should stop after "GA_PostNL.depots".

**Reply by BLindeboom27** [reply]
HI,

Thanks again for the explanation and the fast reply. What you are saying actually makes a lot of sense. The hardcode option for the result function is working, thanks! 

Is the solution you provide possible for both functional and object-oriented programming (in Python)? Meaning that the syntax is the same but for the object-oriented coding you need to specify the class you are using before a function is called.

**Reply by t-wolfeadam** [reply]
My pleasure! And in general, you can think of Pypeline as acting like an interactive Python session ([reference](https://docs.python.org/3/tutorial/interpreter.html#interactive-mode)), where it maintains all variables or functions you create, in addition to any imports, for that "session" (which in this case is a single simulation run). 

To your point about needing to reference the class: no, that's not necessary, I was just following your example code 🙂 . You could do `from GA_PostNL import GAresult, depots, distance_matrix` and then you can reference these variables/functions directly.
